File: lib/network/packets/server/s_attack.dart
/// 攻擊動畫的呈現方式。
enum AttackAnimType {
  /// 直接命中（近身攻擊、瞬發法術）：只在目標身上播 hitGfx。
  direct,

  /// 飛行道具（弓箭、飛劍、火球）：flyGfx 由施放者飛向目標，抵達後播 hitGfx。
  projectile;

  /// 由伺服器代號轉列舉；未知代號一律視為 [direct]。
  static AttackAnimType fromCode(int code) =>
      code == 1 ? AttackAnimType.projectile : AttackAnimType.direct;
}

/// 攻擊演出：只負責演出與傷害數字。
///
/// 攻擊者與目標一律以 objId 識別，玩家、NPC、怪物都適用。
/// 三個動畫編號皆可為 0 表示不播該段。
///
/// 血量不在本封包內 —— 血條由 [SHpUpdate] 獨立更新（血量會因治療、
/// 自然回復等非攻擊原因改變）；目標死亡則由 S_OBJECT_REMOVE 表達。
class SAttack {
  const SAttack({
    required this.attackerObjId,
    required this.targetObjId,
    required this.damage,
    required this.animType,
    required this.castGfx,
    required this.flyGfx,
    required this.hitGfx,
    this.hit = true,
  });

  final int attackerObjId;
  final int targetObjId;
  final int damage;

  final AttackAnimType animType;

  /// 施放者身上的動畫編號（0=無）。
  final int castGfx;

  /// 飛行中的動畫編號（僅 [AttackAnimType.projectile] 使用，0=無）。
  final int flyGfx;

  /// 命中時目標身上的動畫編號（0=無）。
  final int hitGfx;

  /// 是否命中。未命中時 [damage] 為 0，且目標血量不變。
  ///
  /// 不用「damage == 0」判斷 —— 傷害有下限 1，但將來若加入「完全格擋」，
  /// 0 傷害與未命中就不是同一回事了。
  final bool hit;

  factory SAttack.fromData(Map<String, dynamic> data) => SAttack(
        attackerObjId: (data['attackerObjId'] as num?)?.toInt() ?? 0,
        targetObjId: (data['targetObjId'] as num?)?.toInt() ?? 0,
        damage: (data['damage'] as num?)?.toInt() ?? 0,
        animType:
            AttackAnimType.fromCode((data['animType'] as num?)?.toInt() ?? 0),
        castGfx: (data['castGfx'] as num?)?.toInt() ?? 0,
        flyGfx: (data['flyGfx'] as num?)?.toInt() ?? 0,
        hitGfx: (data['hitGfx'] as num?)?.toInt() ?? 0,
        // 舊版伺服器沒有這個欄位，缺少時視為命中
        hit: data['hit'] as bool? ?? true,
      );
}

File: src/main/java/com/xin/server/packet/server/S_Attack.java
package com.xin.server.packet.server;

import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;
import com.xin.server.types.AttackAnimType;

/**
 * 攻擊演出（對應 opcode {@code S_ATTACK}）。
 * <p>
 * 只負責「演出」與「傷害數字」：誰打誰、掉多少血、播哪些動畫。
 * 攻擊者與目標一律以 {@code objId} 識別，玩家、NPC、怪物都適用。
 * <p>
 * <b>血量不在本封包內</b> —— 血條由 {@link S_HpUpdate} 獨立更新，
 * 因為血量會因非攻擊原因（治療、自然回復、buff）而改變；
 * 目標死亡則由 {@link S_ObjectRemove} 表達。
 * <p>
 * 動畫依 {@code animType}（見 {@link AttackAnimType}）分兩種演法：
 * <ul>
 *   <li>{@code 0}=直接命中：只在目標身上播 {@code hitGfx}（近身攻擊、瞬發法術）</li>
 *   <li>{@code 1}=飛行道具：{@code flyGfx} 由施放者飛向目標，抵達後才播 {@code hitGfx}
 *       （弓箭、飛劍、火球）</li>
 * </ul>
 * 三個動畫編號皆可為 {@code 0} 表示不播該段。
 *
 * <pre>
 * {
 *   "op": "S_ATTACK",
 *   "data": {
 *     "attackerObjId": 2000000001, "targetObjId": 1000000005, "damage": 37,
 *     "animType": 1, "castGfx": 120, "flyGfx": 121, "hitGfx": 122
 *   }
 * }
 * </pre>
 */
public class S_Attack extends ServerBasePacket {

    private S_Attack(long attackerObjId, long targetObjId, int damage,
                     int animType, int castGfx, int flyGfx, int hitGfx) {
        super(ServerOpcodes.S_ATTACK);
        put("attackerObjId", attackerObjId);
        put("targetObjId",   targetObjId);
        put("damage",        damage);
        put("animType",      animType);
        put("castGfx",       castGfx);
        put("flyGfx",        flyGfx);
        put("hitGfx",        hitGfx);
        // 明確標示是否命中。不用「damage == 0」判斷 —— 傷害有下限 1，
        // 但將來若加入「完全格擋」之類的機制，0 傷害與未命中就不是同一回事了。
        put("hit",           damage > 0);
    }

    /**
     * 建立攻擊演出封包。
     *
     * @param attackerObjId 攻擊者物件編號
     * @param targetObjId   目標物件編號
     * @param damage        造成的傷害（前端跳的傷害數字）
     * @param animType      動畫呈現方式，見 {@link AttackAnimType}
     * @param castGfx       施放者身上的動畫編號（0=無）
     * @param flyGfx        飛行中的動畫編號（{@code animType=1} 才用，0=無）
     * @param hitGfx        命中時目標身上的動畫編號（0=無）
     */
    /** 未命中：只播動作，不扣血。 */
    public static S_Attack miss(long attackerObjId, long targetObjId) {
        return new S_Attack(attackerObjId, targetObjId, 0,
                AttackAnimType.DIRECT, 0, 0, 0);
    }

    public static S_Attack of(long attackerObjId, long targetObjId, int damage,
                              int animType, int castGfx, int flyGfx, int hitGfx) {
        return new S_Attack(attackerObjId, targetObjId, damage,
                animType, castGfx, flyGfx, hitGfx);
    }

    /**
     * 建立直接命中（{@link AttackAnimType#DIRECT}）的攻擊演出封包，
     * 只在目標身上播一段命中動畫。
     */
    public static S_Attack direct(long attackerObjId, long targetObjId, int damage, int hitGfx) {
        return new S_Attack(attackerObjId, targetObjId, damage,
                AttackAnimType.DIRECT, 0, 0, hitGfx);
    }
}
